src/arena.rs

```diff
@@ -74,10 +74,7 @@ impl RawBlockTraits for F64Table {
         F64_TABLE_INIT_SIZE
     }
 
-    #[inline]
-    fn align() -> usize {
-        F64_TABLE_ALIGN
-    }
+    const ALIGN: usize = F64_TABLE_ALIGN;
```

Comment:
Would it maybe make sense for `RawBlockTraits` to have an associated type `Content` instead which would then be used via [`std::mem::align_of`](https://doc.rust-lang.org/std/mem/fn.align_of.html) were necessary to get the alignment rather than having some alignment constant which we then need set correctly manually and to ensure correctness assert to be equal to some types alignment anyway?

Reply:
For `F64Table` this appears to be `UnsafeCell<OrderedFloat<f64>>`, for `AtomTable` this looks to be `AtomData`, and for `Stack` we would probably want a new union  type of `AndFrame`, `OrFrame`, and `HeapCellValue`.

Reply:
Well, `AtomTable` stores an `AtomData`, which is unsized (and whose `align_of` cannot be taken because of compiler limitations); should `AtomTable::Content` then be `AtomHeader`?

Reply:
Ah, looks like I missed that detail.

Your suggestion should work as the `data` field in `AtomData` shouldn't increase the alignment as `str` should have an alignment of 1. For now I would consider it a sufficiently good workaround.
If in the future https://github.com/rust-lang/rfcs/pull/3319 gets accepted and implemented we could get rid of the workaround, maybe include a comment mentioning that.

Reply:
There is also [`Layout::align`](https://doc.rust-lang.org/stable/std/alloc/struct.Layout.html#method.align) which does not have a sized bound but to use it we would need a reference to a value so we can construct a `Layout` instance via [`Layout::for_val`](https://doc.rust-lang.org/stable/std/alloc/struct.Layout.html#method.for_value) so we would be in a chicken before the egg scenario.

Reply:
Hmm, I don't know about this change, it looks hacky: the `RawBlock` doesn't directly store `AtomHeader`s or a union of `AndFrame` and `OrFrame`. It stores values of variable length, encoded in a custom format each time: `AtomHeader` followed by `str`, `AndFrame` followed by the `HeapCellValue`.

(`F64Table` could really just be a `Vec<UnsafeCell<OrderedFloat<f64>>>`)

I think that if I make this change, then it would be by fully changing the name and API of `RawBlock` to fit this "fixed-size header followed by a variable number of values" model.

That's certainly doable and a much cleaner solution than the current one, but I fear it would be too much for one PR.

Reply:
I'm moving this idea to a separate issue then: #2737